activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, **options)
+      raise ArgumentError, "inserts should be an array" unless inserts.is_a?(Array)
```

Comment:
You're only using inserts as an argument for the #prepare_keys_and_values_for_insert, so you can push the arg validation in there, after you've pulled the method over into this class.

Reply:
When I pulled `prepare_keys_and_values_for_insert` into this class, it broke apart into two methods, `keys`and `values_list` (and later, I pushed `values_list` down to `InsertAll::Builder`). Now that this object hangs on to `inserts` longer, I _think_ it makes sense to leave the validation here. But another logical place to put it would be the `keys` method on this class. Do you think it should go there?

Reply:
Remind me again with you need this protection? When wouldn't inserts not be an array?

Reply:
> When wouldn't inserts not be an array?

Only if a user passed something goofy to `insert_all` like:

```ruby
Book.insert_all! 5
```

The trade off is whether they get:

```
ArgumentError, "inserts should be an array"
```

or

```
NoMethodError, "undefined method `first' for 5:Integer"
```

I was trying to give a clearer and more useful error message; but it doesn't some to be _lots_ better. Do you want me to cut it?

Reply:
Yeah, not sure that's worth it. Just documenting the call signature should be fine.

Reply:
✅ **removed**